packages/nextjs/src/config/wrappers/withSentryAPI.ts

```diff
@@ -189,7 +176,8 @@ export function withSentry(origHandler: NextApiHandler, parameterizedRoute?: str
         // out. (Apps which are deployed on Vercel run their API routes in lambdas, and those lambdas will shut down the
         // moment they detect an error, so it's important to get this done before rethrowing the error. Apps not
         // deployed serverlessly will run into this cleanup function again in `res.end(), but it'll just no-op.)
-        await finishSentryProcessing(res);
+        await finishTransaction(res);
+        await flushQueue();
```

Comment:
Aren't we calling these two functions twice now for tha API routes? Once here and once implicitly above with `autoEndTransactionOnResponseEnd`? If so, we could delete `finishTransaction` by moving it's logic into `autoEndTransactionOnResponseEnd` which would also remove the need for `__sentryTransaction`.

Reply:
The problem is that `res.end()` doesn't always get called if there's an error. See https://github.com/getsentry/sentry-javascript/pull/4027.

Reply:
Ah ok got it. If that is the case, I find the `__sentryTransaction` solution still a little bit overcomplicated. I think we should just lift up the declaration of the `transaction` variable and pass in the transaction directly instead of writing it on `res`.

Something like (this would also ofc involve changing the signature of `finishTransaction`):

```diff
@@ -1,5 +1,6 @@
 import { captureException, getCurrentHub, startTransaction } from '@sentry/node';
 import { extractTraceparentData, hasTracingEnabled } from '@sentry/tracing';
+import { Transaction } from '@sentry/types';
 import {
   addExceptionMechanism,
   baggageHeaderToDynamicSamplingContext,
@@ -76,6 +77,7 @@ export function withSentry(origHandler: NextApiHandler, parameterizedRoute?: str
     // getting that before it will finish the response.
     const boundHandler = local.bind(async () => {
       const currentScope = getCurrentHub().getScope();
+      let transaction: Transaction | undefined;
 
       if (currentScope) {
         currentScope.setSDKProcessingMetadata({ request: req });
@@ -110,7 +112,7 @@ export function withSentry(origHandler: NextApiHandler, parameterizedRoute?: str
 
           const reqMethod = `${(req.method || 'GET').toUpperCase()} `;
 
-          const transaction = startTransaction(
+          transaction = startTransaction(
             {
               name: `${reqMethod}${reqPath}`,
               op: 'http.server',
@@ -176,7 +178,7 @@ export function withSentry(origHandler: NextApiHandler, parameterizedRoute?: str
         // out. (Apps which are deployed on Vercel run their API routes in lambdas, and those lambdas will shut down the
         // moment they detect an error, so it's important to get this done before rethrowing the error. Apps not
         // deployed serverlessly will run into this cleanup function again in `res.end(), but it'll just no-op.)
-        await finishTransaction(res);
+        await finishTransaction(transaction);
         await flushQueue();
 
         // We rethrow here so that nextjs can do with the error whatever it would normally do. (Sometimes "whatever it
```

Wdyt?

Reply:
I don't feel strongly either way, but I'll admit I don't get how attaching the transaction to the response is any more complicated than attaching it to the request (which we do elsewhere and which was, IIRC, your idea...)

Reply:
> I don't get how attaching the transaction to the response is any more complicated than attaching it to the request (which we do elsewhere and which was, IIRC, your idea...)

Attaching the transaction to the request was done out of necessity since datafetchers need to share a transaction to spawn child spans, but they have no other reliable way of sharing data except via the req/res objects.

I would never do that simply because it's an option. Attaching stuff to req/res, just to extract it in the step directly after (while we would have it available as a simple variable) is just an indirection I'd like to avoid.

Reply:
But we can do this in a separate PR.